pkg/controllers/devicetemplaterevision/controller.go

```diff
@@ -0,0 +1,136 @@
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	name = "device-template-revision-controller"
+)
+
+const (
+	templateRevisionReference = "edgeapi.cattle.io/template-revision-reference"
+	deviceTemplateKindName    = "DeviceTemplate"
+)
+
+type Controller struct {
+	context            context.Context
+	templateController controllers.DeviceTemplateController
+	templateLister     controllers.DeviceTemplateCache
+	revisionController controllers.DeviceTemplateRevisionController
+	revisionLister     controllers.DeviceTemplateRevisionCache
+	apply              apply.Apply
+}
+
+func Register(ctx context.Context, apply apply.Apply, revisionController controllers.DeviceTemplateRevisionController, templateController controllers.DeviceTemplateController) {
+	ctrl := &Controller{
+		context:            ctx,
+		templateController: templateController,
+		templateLister:     templateController.Cache(),
+		revisionController: revisionController,
+		revisionLister:     revisionController.Cache(),
+		apply:              apply,
+	}
+	revisionController.OnChange(ctx, name, ctrl.OnChanged)
+	revisionController.OnRemove(ctx, name, ctrl.OnRemoved)
+}
+
+func (c *Controller) OnChanged(key string, obj *v1alpha1.DeviceTemplateRevision) (*v1alpha1.DeviceTemplateRevision, error) {
+	if key == "" {
+		return nil, nil
+	}
+
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return nil, nil
+	}
+	deviceTemplate, err := c.templateLister.Get(obj.Namespace, obj.Spec.DeviceTemplateName)
+	if err != nil {
+		return nil, err
+	}
+
+	objCopy := obj.DeepCopy()
+	objCopy.Status.UpdatedAt = metav1.Time{Time: time.Now()}
+	objCopy.OwnerReferences = append(objCopy.OwnerReferences[:0], SetRevisionOwner(objCopy, deviceTemplate.UID))
+
+	if err := c.SyncDeviceTemplateDefaultRevision(objCopy, deviceTemplate, false); err != nil {
+		return nil, err
+	}
+
+	return c.revisionController.Update(objCopy)
+}
+
+func (c *Controller) OnRemoved(key string, obj *v1alpha1.DeviceTemplateRevision) (*v1alpha1.DeviceTemplateRevision, error) {
+	if key == "" {
+		return obj, nil
+	}
+
+	deviceTemplate, err := c.templateLister.Get(obj.Namespace, obj.Spec.DeviceTemplateName)
+	if err != nil {
+		if !apierrs.IsNotFound(err) {
+			return nil, err
+		}
+		return nil, nil
+	}
+	if err := c.SyncDeviceTemplateDefaultRevision(obj, deviceTemplate, true); err != nil {
+		return nil, err
+	}
+
+	return obj, c.revisionController.Delete(obj.Namespace, obj.Name, &metav1.DeleteOptions{})
+}
+
+func (c *Controller) SyncDeviceTemplateDefaultRevision(obj *v1alpha1.DeviceTemplateRevision, deviceTemplate *v1alpha1.DeviceTemplate, onRemove bool) error {
+	set := labels.Set(map[string]string{templateRevisionReference: obj.Spec.DeviceTemplateName})
+	revisions, err := c.revisionLister.List(obj.Namespace, set.AsSelector())
+	if err != nil {
+		return err
+	}
+	revisionCount := len(revisions)
+	if onRemove {
+		if revisionCount == 1 && deviceTemplate.Spec.DefaultRevisionName == obj.Namespace+"/"+revisions[0].Name {
```

Comment:
assign the `obj.Namespace+"/"+revisions[0].Name` on the above with `fmt.Sprintf`